### ISSUE TITLE
Shader compiler errors on RPi-4B, only the overlay renders correctly

### ISSUE THREAD

**Issue by amverweij** [hrydgard/ppsspp#16824]
### Game or games this happens in

ULUS10457 : SOULCALIBUR: Broken Destiny

### What area of the game / PPSSPP

In Training mode or in Quick Match

### What should happen

The scene should be rendered correctly.

### Logs

[retroarch.log](https://github.com/hrydgard/ppsspp/files/10472011/retroarch.log)

### Platform

Raspberry Pi

### Mobile phone model or graphics card

Broadcom VC4/V3D

### PPSSPP version affected

03f64a021651b908bb1ad912862525c10f0adef3

### Last working version

a73ccd77e327600470b806341ab6f81177b37b5b, but I can narrow it down further

### Graphics backend (3D API)

Vulkan

### Checklist

- [X] Test in the [latest git build](https://buildbot.orphis.net/ppsspp/) in case it's already fixed.
- [X] [Search for other reports](https://github.com/hrydgard/ppsspp/search?q=my+issue&type=issues) of the same issue.
- [X] Try resetting settings or older versions and include if the issue is related.
- [X] Try without any cheats and without loading any save states.
- [X] Include logs or screenshots of issue.

**Comment by amverweij** [comment]
I was curious about your recent Vulkan improvements. (Hoping that one day you manage to get critical finishes in Soulcalibur Broken Destiny work too on my RPi-4B.)

**Comment by amverweij** [comment]
I was curious about your recent Vulkan improvements. (Hoping that one day you manage to get critical finishes in Soulcalibur Broken Destiny work too on my RPi-4B.)

**Comment by amverweij** [comment]
In case you want to take this up, I can do a git bisect to identify the commit that broke it.

**Comment by amverweij** [comment]
Here is a video of the issue made with my mobile phone:

https://user-images.githubusercontent.com/51815579/213864346-f8867f37-2a8a-45e2-b33d-22c926b2fec9.mp4

But I guess the log file tells enough.

**Comment by amverweij** [comment]
I'm using Libretro.

**Comment by hrydgard** [comment]
Thanks for reporting! A git bisect would speed things up indeed.

**Comment by amverweij** [comment]
Not sure I can run PPSSPP standalone, I'm on KMS/DRM. My understanding was that is still  [pending](https://github.com/hrydgard/ppsspp/pull/14758). I will do the bisect and report back later.

**Comment by amverweij** [comment]
Git-bisect indicates the first bad commit is 3b4640935025fbd478d708811402944132e87a60.

**Comment by hrydgard** [comment]
OK, that's not entirely surprising, but a bit disappointing that such a simple thing would fail.

I guess we have to revert to the previous behavior. Hopefully it will be enough to turn off the loop on Broadcom, so we don't have to have two buffer types...

**Comment by hrydgard** [comment]
#17222 will hopefully have fixed it, if not, due to this being a rather bad driver bug that will hopefully get fixed in the near future if it isn't already, I'm not really willing to do more work on this for 1.15, so moving the milestone.

**Comment by amverweij** [comment]
I will test and report back. Also, I'm on a newer version of mesa then when I logged the bug, so who knows.

**Comment by amverweij** [comment]
Hmm, ppsspp git master compiles against glslang 12, however I didn't adapt the gentoo live ebuild for this yet. More later, TBC.

**Comment by amverweij** [comment]
https://github.com/hrydgard/ppsspp/commit/9f6338e0981e6843f7bcf1ab7592a22cf4795a2d renders correctly on my RPi-4B.

**Comment by unknownbrackets** [comment]
Thanks for confirming.

-[Unknown]

**Comment by amverweij** [comment]
I also tested bb351e051167072e93892f4bfa06f690626c120c, and that one also renders correctly. 

This means that #17222 wasn't actually fixing anything. Something else (either in ppsspp or in the newer mesa) has fixed the issue.

**Comment by hrydgard** [comment]
Hm, okay. Well, at least now we're compatible backwards to some older driver version.

Can you check in Settings/Tools/System Information what the version of the driver is, so we can make the workaround conditional?

**Comment by amverweij** [comment]
Sure, I love to help. Could be later this week though, compiling on the RPi4-B is slow and I will need to first reproduce the issue and then see what fixed it. So that will be some work.

**Comment by hrydgard** [comment]
No, I'm asking for something much simpler. Just start it up and go and check what the driver version is reported as.

We can assume that we don't need the workaround for sure from that version and beyond, and that's good enough.

**Comment by amverweij** [comment]
I'm currently on mesa-23.0.0.

I'm running the libretro core from retroarch. There doesn't seem to be a reference to the driver version in the GUI, strange as that may sound.  

Here is the Video part of the log file, in case that might help:

[INFO] [Video]: Set video size to: 1905x1080.
[INFO] [Video]: Using HW render, vulkan driver forced.
[INFO] [Vulkan]: Vulkan dynamic library loaded.
[INFO] [Vulkan]: Found vulkan context: "khr_display".
[INFO] [Vulkan]: Detecting screen resolution: 0x0.
[INFO] [Vulkan]: Found GPU at index 0: "V3D 4.2".
[INFO] [Vulkan]: Using GPU index 0.
[INFO] [Vulkan]: Got HW context negotiation interface 1.
[INFO] [Vulkan]: Found GPU at index 0: "V3D 4.2".
[INFO] [Vulkan]: Using GPU index 0.
[INFO] [Vulkan]: Using semaphores for WSI acquire.
[INFO] [Vulkan]: Using GPU: "V3D 4.2".
[INFO] [Vulkan]: Got 3 swapchain images.
[INFO] [Vulkan]: Using resolution 1920x1080.
[INFO] [Vulkan]: Using BGRA8888 format.
[INFO] [Vulkan]: Loading stock shader.
[INFO] [Video]: Graphics driver did not initialize an input driver. Attempting to pick a suitable driver.

**Comment by amverweij** [comment]
vulkaninfo reports:

VkPhysicalDeviceDriverProperties:
---------------------------------
        driverID        = DRIVER_ID_MESA_V3DV
        driverName      = V3DV Mesa
        driverInfo      = Mesa 23.0.0
        conformanceVersion:
                major    = 1
                minor    = 2
                subminor = 7
                patch    = 1

**Comment by hrydgard** [comment]
Really? You should see a driver version if you follow Settings/Tools/System Information and scroll down a bit. Like this from my mac, where I'd want to know 0.2.194 ( 0000277a ):

<img width="940" alt="image" src="https://user-images.githubusercontent.com/130929/229499590-9bf30c02-2a38-44d5-bb2a-000e1623c0dc.png">

**Comment by amverweij** [comment]
Your screenshot is from PPSSPP standalone. I'm running RetroArch on KMS/GBM/EGL.

Last time that I tried I wasn't able to run PPSSPP standalone on KMS/GBM/EGL. As far as I can see, running PPSSPP standalone requires me to install a complete desktop environment.

**Comment by hrydgard** [comment]
Okay. Well, what you got above might be enough actually. Thanks!

**Comment by amverweij** [comment]
I have been able to reproduce the issue again with https://github.com/hrydgard/ppsspp/commit/03f64a021651b908bb1ad912862525c10f0adef3, but only after switching to using system glslang. This used to be the way PPSSPP and its Libretro core were compiled on Gentoo Linux.

It follows that this was not a driver issue, but an incompatibility between the shared glslang libraries installed on Gentoo Linux (currently glslang-1.3.224-r1) and the one PPSSPP uses during its build process.

The ppsspp ebuild in the main Gentoo repository switched from using system glslang to using internal glslang for PPSSPP  [on 2023-03-04](https://gitweb.gentoo.org/repo/gentoo.git/commit/games-emulation/ppsspp?id=d944d80cea26de120e9e491ca344638cdc616a8f).

The ppsspp-libretro ebuild followed [on 2023-03-23](https://gitlab.com/menelkir/gentoo-overlay/-/commit/a400031f799e3a194c2e4381418742fc9e20d689).

So it seems you can safely revert https://github.com/hrydgard/ppsspp/pull/17222. It is not needed for fixing any rendering issues.

Sorry for the inconvenience. And thank you for making this great emulator!

**Comment by hrydgard** [comment]
Cool, thank you for tracking that down!

I'll revert the commit.
